### PR TITLE
Add bfloat16 support to binary_cross_entropy for CPU

### LINKED ISSUE #123715: Add support to bf16 element type for binary_cross_entropy.
### 🚀 The feature, motivation and pitch

Please add support to bf16 element type for binary_cross_entropy.

### Alternatives

_No response_

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by fffrog**
Fixes #123715

As the title stated.

But, maybe we should pay attention to this https://github.com/pytorch/pytorch/pull/33206, which removed the half support for cpu about 4 years ago.

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -273,27 +273,26 @@ Tensor& binary_cross_entropy_out_cpu(const Tensor& input, const Tensor& target,
       .add_owned_const_input(at::squeeze(target))
       .build();
 
-    AT_DISPATCH_FLOATING_TYPES(loss.scalar_type(), "binary_cross_entropy", [&] {
-        at::native::cpu_kernel(
-            iter,
-            [] (scalar_t input_val, scalar_t target_val) {
+    AT_DISPATCH_FLOATING_TYPES_AND(
+        ScalarType::BFloat16, loss.scalar_type(), "binary_cross_entropy", [&] {
```
Comment (Skylion007) [inline aten/src/ATen/native/Loss.cpp:277]: Any reason we don't add fp16 while we are at it? numerical reasons?

Comment (FFFrog) [reply aten/src/ATen/native/Loss.cpp:277]: Sorry, please look at https://github.com/pytorch/pytorch/pull/33206, i am not sure whether the performance of fp16 for CPU is been fix or not.

Comment (FFFrog) [reply aten/src/ATen/native/Loss.cpp]: Sorry to bother @Skylion007 again.

I have dived into log1p and log, which make up the binary_cross_entropy operator, and I found that vectorization of them was already supported, so I added "Half" support for binary_cross_entropy on the CPU.

**Review by Skylion007 — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Review by malfet — COMMENTED** [review]
Should we create some sort of a new issues, as a lot of things has happened since 4 years ago, for example CPU no longer equals x86_64 only, and on ARM half support makes a lot of sense. But there are also precision consideration

**Comment by malfet** [comment]
@FFFrog looks you've uncovered a bug in MPS binary cross-entropy implementation
```
test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_binary_cross_entropy_cpu_float16 loc("mps_subtract"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":228:0)): error: input types 'tensor<f32>' and 'tensor<1xf16>' are not broadcast compatible
```
Please file an issue and disable the test for this dtype (there is a skip list)

**Comment by FFFrog** [comment]
> @FFFrog looks you've uncovered a bug in MPS binary cross-entropy implementation
> 
> ```
> test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_binary_cross_entropy_cpu_float16 loc("mps_subtract"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":228:0)): error: input types 'tensor<f32>' and 'tensor<1xf16>' are not broadcast compatible
> ```
> 
> Please file an issue and disable the test for this dtype (there is a skip list)

Thank you for your reply. I will file an issue about this.
